### PR TITLE
[NEXUS-8860] Browse Assets UI 

### PR DESCRIPTION
**Pull request by adreghiciu**
https://issues.sonatype.org/browse/NEXUS-8860

The PR adds a new feature (menu entry) for browsing assets along side with the existing browse components feature. It does not, as described in issue, add a switch between assets/component after selecting the repository, as that will overcomplicate the drill down implementation.

### CODE REVIEW

**Comment by mcculls** [comment]
will there be functional tests for this feature?

**Comment by djsauble** [comment]
This is awesome. Nice work.

I don’t much care for the current icon though. It’s kind of generic, and fades into the white background.

assets - sonatype nexus 2015-07-10 10-55-49

Maybe the script_binary.png icon instead?

script_binary

**Comment by djsauble** [comment]
Seeing an error when I drill into the asset detail view. I’m running Nexus Pro.

screen shot 2015-07-10 at 11 02 18 am

**Comment by djsauble** [comment]
I also think the separate feature is fine for now. If people need a more seamless interaction between the asset- and component-first views, we’ll learn that soon enough. Not worth investing a bunch of time into a new pattern or extending the drilldown pattern, for now at least.

**Comment by adreghiciu** [comment]
@djsauble probably you did not check out the branch with same name from pro

**Comment by adreghiciu** [comment]
@djsauble can you add the image to this branch so I can use it?

**Comment by djsauble** [comment]
@adreghiciu Done.

**Comment by djsauble** [comment]
Yep, that was my problem. Checked out NEXUS-8860-browseui-asset on the pro branch, and it all works beautifully.

File: plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+   * When a list managed by this controller is clicked, route the event to the proper handler
+   */
+  onSelection: function(list, model) {
+    var me = this,
+        modelType;
+
+    // Figure out what kind of list we’re dealing with
+    modelType = model.id.replace(/^.*?model\./, '').replace(/\-.*$/, '');
+
+    if (modelType == "RepositoryReference") {
```
Comment (sonatype-zion) [inline plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js]: MAJOR Replace "==" with "===". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AEqEqEq)

File: plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    'Asset'
+  ],
+
+  views: [
+    'browse.BrowseAssetFeature',
+    'browse.BrowseAssetList',
+    'browse.BrowseRepositoryList'
+  ],
+
+  refs: [
```
Comment (sonatype-zion) [inline plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js]: BLOCKER Avoid trailing comma in array and object literals. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ATrailingComma)

File: plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginStrings.js [hunk 3]
```diff
@@ -32,19 +32,23 @@ Ext.define('NX.coreui.app.PluginStrings', {
 
   keys: {
```
Comment (sonatype-zion) [inline plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginStrings.js]: BLOCKER Avoid trailing comma in array and object literals. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ATrailingComma)

**Comment by sonatype-zion** [comment]
SonarQube analysis reported 22 issues:
* BLOCKER 2 blocker
* MAJOR 12 major
* MINOR 7 minor
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Sections of code should not be "commented out".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginConfig.js#L78) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ACommentedCode)
* MAJOR [Sections of code should not be "commented out".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginConfig.js#L76) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ACommentedCode)
* MAJOR [Sections of code should not be "commented out".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginConfig.js#L80) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ACommentedCode)
* MAJOR [Sections of code should not be "commented out".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginConfig.js#L79) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ACommentedCode)
* MAJOR [Replace "==" with "===".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseComponents.js#L120) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AEqEqEq)
* MAJOR [Replace "==" with "===".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseComponents.js#L117) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AEqEqEq)
* MAJOR [Take the required action to fix the issue indicated by this comment.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/HealthCheckColumns.js#L293) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AS1134)
* MAJOR [Function has a complexity of 11 which is greater than 10 authorized.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/HealthCheckColumns.js#L290) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AFunctionComplexity)
* MAJOR [Replace "!=" with "!==".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/Search.js#L552) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AEqEqEq)
* MAJOR [Replace "!=" with "!==".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/Search.js#L274) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AEqEqEq)
* INFO [Complete the task associated to this TODO comment.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/Search.js#L445) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AS1135)
* MINOR [Add a semicolon at the end of this statement.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/Search.js#L171) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ASemicolon)
* MAJOR [Replace "==" with "===".](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/Search.js#L474) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AEqEqEq)
* MINOR [Add a semicolon at the end of this statement.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/Search.js#L553) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ASemicolon)
* MINOR [Add a semicolon at the end of this statement.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/view/component/AssetAttributes.js#L86) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ASemicolon)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/view/component/AssetInfo.js#L29) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/view/component/AssetInfo.js#L32) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/view/component/AssetInfo.js#L54) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/sonatype/nexus-oss/blob/cfedcdaa34f6137dfb15c86195af84b832f11785/plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/view/component/AssetInfo.js#L66) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js (hunk 2 above)
Comment (adreghiciu) [inline plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js]: Fixed

File: plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginStrings.js (hunk 3 above)
Comment (adreghiciu) [inline plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/app/PluginStrings.js]: Fixed

File: plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js (hunk 1 above)
Comment (adreghiciu) [inline plugins/rapture/nexus-coreui-plugin/src/main/resources/static/rapture/NX/coreui/controller/BrowseAssets.js]: Fixed

**Comment by mcculls** [comment]
Seeing the following oddities when testing this locally:
* browser back/forward buttons no longer affect the UI
* drilling-down to an asset (such as asm-3.1.pom downloaded from maven-central repo via curl) shows an error in the component info panel: "ERROR: (ID ...) java.lang.IllegalArgumentException: A segment is null"
* stack trace shows:
```
java.lang.IllegalArgumentException: A segment is null
	at org.jboss.resteasy.specimpl.ResteasyUriBuilder.segment(ResteasyUriBuilder.java:983) [org.sonatype.siesta.server:2.2.0.SNAPSHOT]
	at com.sonatype.nexus.plugins.healthcheck.service.impl.UrlServiceImpl.buildInsightUrl(UrlServiceImpl.java:103) [na:na]
	at com.sonatype.nexus.plugins.healthcheck.service.impl.UrlServiceImpl.getComponentDetailsUrl(UrlServiceImpl.java:96) [na:na]
	at com.sonatype.nexus.plugins.healthcheck.service.InsightServiceImpl.getComponentDetails(InsightServiceImpl.java:107) [na:na]
	at com.sonatype.nexus.plugins.healthcheck.pro.rest.HealthCheckFileDetailResource.getContent(HealthCheckFileDetailResource.java:245) [na:na]
	at com.sonatype.nexus.plugins.healthcheck.pro.rest.HealthCheckFileDetailResource.doGet(HealthCheckFileDetailResource.java:181) [na:na]
	at com.sonatype.nexus.plugins.healthcheck.pro.rest.HealthCheckFileDetailResource.getComponentDetails(HealthCheckFileDetailResource.java:114) [na:na]
```
* drilling-down to an asset in the component view then clicking on the asset view (ie. 'Assets' in the leftmost column) causes a javascript error: Uncaught TypeError: Cannot read property 'getStore' of undefined
* refreshing the page when viewing an asset moves the UI back to the top-level, but the URL remains the same - suggesting that the bookmarking is not quite right

**Comment by mcculls** [comment]
Note: the browser back/forward issue also happens on master, so I suspect there's a general problem with drill-down atm.

**Comment by djsauble** [comment]
https://github.com/sonatype/nexus-oss/pull/1547 fixes https://issues.sonatype.org/browse/NEXUS-8985

**Comment by adreghiciu** [comment]
* browser back/forward is not related indeed
* segment error is a known issue and is due to commented out code in health check/clm from when we removed legacy repositories https://issues.sonatype.org/browse/NEXUS-8936
* debugged the java script error and is an unrelated drill down issue that will happen always (in certain conditions) when switching from a drill down with more masters to one with less masters (e.g. browse components has 3 and browse assets has 2 masters). The exception happen because drill down controller uses the index  from one drill down (= 3 in example above) to access an array from the other one (which only has 2 entries in array)
* navigation on refresh is also a known issue as in browse components UI. I suggest we solve them both at once as the solution will be the same

**Comment by mcculls** [comment]
+1 given that the basic functionality works, and the observed errors are addressed in other PRs

**Comment by adreghiciu** [comment]
Manually merged https://github.com/sonatype/nexus-oss/commit/a692b4705aeaa26a24d970f646afc47a1fdabdf4

**Pull request closed without merging**